async_substrate_interface/substrate_addons.py

```diff
@@ -0,0 +1,315 @@
+    "decode_scale",
+    "encode_scale",
+    "generate_signature_payload",
+    "get_account_next_index",
+    "get_account_nonce",
+    "get_block",
+    "get_block_hash",
+    "get_block_header",
+    "get_block_metadata",
+    "get_block_number",
+    "get_block_runtime_info",
+    "get_block_runtime_version_for",
+    "get_chain_finalised_head",
+    "get_chain_head",
+    "get_constant",
+    "get_events",
+    "get_extrinsics",
+    "get_metadata_call_function",
+    "get_metadata_constant",
+    "get_metadata_error",
+    "get_metadata_errors",
+    "get_metadata_module",
+    "get_metadata_modules",
+    "get_metadata_runtime_call_function",
+    "get_metadata_runtime_call_functions",
+    "get_metadata_storage_function",
+    "get_metadata_storage_functions",
+    "get_parent_block_hash",
+    "get_payment_info",
+    "get_storage_item",
+    "get_type_definition",
+    "get_type_registry",
+    "init_runtime",
+    "initialize",
+    "query",
+    "query_map",
+    "query_multi",
+    "query_multiple",
+    "retrieve_extrinsic_by_identifier",
+    "rpc_request",
+    "runtime_call",
+    "submit_extrinsic",
+    "subscribe_block_headers",
+    "supports_rpc_method",
+]
+
+RETRY_PROPS = ["properties", "version", "token_decimals", "token_symbol", "name"]
+
+
+class RetrySyncSubstrate(SubstrateInterface):
+    """
+    A subclass of SubstrateInterface that allows for handling chain failures by using backup chains. If a sustained
+    network failure is encountered on a chain endpoint, the object will initialize a new connection on the next chain in
+    the `fallback_chains` list. If the `retry_forever` flag is set, upon reaching the last chain in `fallback_chains`,
+    the connection will attempt to iterate over the list (starting with `url`) again.
+
+    E.g.
+    ```
+    substrate = RetrySyncSubstrate(
+        "wss://entrypoint-finney.opentensor.ai:443",
+        fallback_chains=["ws://127.0.0.1:9946"]
+    )
+    ```
+    In this case, if there is a failure on entrypoint-finney, the connection will next attempt to hit localhost. If this
+    also fails, a `MaxRetriesExceeded` exception will be raised.
+
+    ```
+    substrate = RetrySyncSubstrate(
+        "wss://entrypoint-finney.opentensor.ai:443",
+        fallback_chains=["ws://127.0.0.1:9946"],
+        retry_forever=True
+    )
+    ```
+    In this case, rather than a MaxRetriesExceeded exception being raised upon failure of the second chain (localhost),
+    the object will again being to initialize a new connection on entrypoint-finney, and then localhost, and so on and
+    so forth.
+    """
+
+    def __init__(
+        self,
+        url: str,
+        use_remote_preset: bool = False,
+        fallback_chains: Optional[list[str]] = None,
+        retry_forever: bool = False,
+        ss58_format: Optional[int] = None,
+        type_registry: Optional[dict] = None,
+        type_registry_preset: Optional[str] = None,
+        chain_name: str = "",
```

Comment:
```suggestion
        chain_name: str = "Bittensor",
```
just as default. the same for second class

Reply:
We shouldn't do this, because this isn't just designed for Bittensor. Same reason we don't use this default in AsyncSubstrateInterface or SubstrateInterface.